File: ruby/specs/02_run_testsets.rb
# -*- coding: utf-8 -*-

$LOAD_PATH.push '../lib' unless $LOAD_PATH.include?('../lib')

require 'yaml'
require 'woothee'

TESTSET_DIR = File.dirname(__FILE__) + "/../../testsets/";

TARGETS = [
#           ['crawler.yaml','Crawler'],['crawler_google.yaml','Crawler/Google'],
           ['pc_windows.yaml','PC/Windows'],#['pc_misc.yaml','PC/Misc'],
#           ['mobilephone_docomo.yaml','MobilePhone/docomo'],['mobilephone_au.yaml','MobilePhone/au'],
#           ['mobilephone_softbank.yaml','MobilePhone/softbank'],['mobilephone_willcom.yaml','MobilePhone/willcom'],
#           ['mobilephone_misc.yaml','MobilePhone/misc'],
#           ['smartphone_ios.yaml','SmartPhone/ios'],['smartphone_android.yaml','SmartPhone/android'],
#           ['smartphone_misc.yaml','SmartPhone/misc'],
#           ['appliance.yaml','Appliance'],
#           ['pc_lowpriority.yaml','PC/LowPriority'],
#           ['misc.yaml','Misc'],
#           ['crawler_nonmajor.yaml','Crawler/NonMajor'],
          ]

describe Woothee do
  
  TARGETS.each do |filename,groupname|
    YAML.load_file(TESTSET_DIR + filename).each do |e|
      r = Woothee.parse(e['target'])
      [:name, :category, :os, :version, :vendor].each do |attribute|
        it groupname + ("test(%s): %s" % [attribute, e['target']]) do
          if [:name, :category, :os].include?(attribute) or ([:version, :vendor].include?(attribute) and e.has_key?(attribute.to_s))
            r[attribute].to_s.should eql(e[attribute.to_s])
          end
        end
      end
    end
    # it groupname do 
    #   YAML.load_file(TESTSET_DIR + filename).each do |e|
    #     # TODO we need contents of 'r' and 'e', instead of rspec code itself
    #     r = Woothee.parse(e['target'])
    #     r.should_not be_nil
    #     r[:name].should eql(e['name'])
    #     r[:category].to_s.should eql(e['category'])
    #     r[:os].should eql(e['os']) if e.has_key?('os')
    #     r[:version].should eql(e['version']) if e.has_key?('version')
    #     r[:vendor].should eql(e['vendor']) if e.has_key?('vendor')
    #   end
    # end
  end

end

File: js/test/02_run_testsets.js
var expect = require('chai').expect;

require('js-yaml');

var woothee = require('../lib/woothee'),
    dataset = require('../lib/dataset');

var TESTSET_DIR = process.cwd() + '/../testsets/',
    TARGETS = [
      ['crawler.yaml','Crawler'],['crawler_google.yaml','Crawler/Google'],
      ['pc_windows.yaml','PC/Windows'],['pc_misc.yaml','PC/Misc'],
      ['mobilephone_docomo.yaml','MobilePhone/docomo'],['mobilephone_au.yaml','MobilePhone/au'],
      ['mobilephone_softbank.yaml','MobilePhone/softbank'],['mobilephone_willcom.yaml','MobilePhone/willcom'],
      ['mobilephone_misc.yaml','MobilePhone/misc'],
      ['smartphone_ios.yaml','SmartPhone/ios'],['smartphone_android.yaml','SmartPhone/android'],
      ['smartphone_misc.yaml','SmartPhone/misc'],
      ['appliance.yaml','Appliance'],
      ['pc_lowpriority.yaml','PC/LowPriority'],
      ['misc.yaml','Misc'],
      ['crawler_nonmajor.yaml','Crawler/NonMajor']
    ];

describe('woothee', function(){
  TARGETS.forEach(function(pair){
    var filename = pair[0],
        groupname = pair[1];
    var testset = require(TESTSET_DIR + filename);

    testset.forEach(function(entry){

      var result = woothee.parse(entry.target);

      dataset.ATTRIBUTE_LIST.forEach(function(attr){
        it(groupname + 'test(' + attr + '): ' + entry.target, function(){

          switch (attr) {
          case dataset.ATTRIBUTE_NAME:
          case dataset.ATTRIBUTE_CATEGORY:
            break;
          case dataset.ATTRIBUTE_OS:
          case dataset.ATTRIBUTE_VERSION:
          case dataset.ATTRIBUTE_VENDOR:
            if (! entry[attr]) return;
            break;
          default:
            throw 'unknown attribute: ' + attr;
          }

          expect(result[attr]).to.equal(entry[attr]);
        });
      });
    });
  });
});
